crates/bevy_geometry/src/lib.rs

```diff
@@ -0,0 +1,359 @@
+                return false;
+            }
+        }
+        true
+    }
+}
+impl AxisAlignedBox {
+    /// An ordered list of the vertices that form the 8 corners of the [AxisAlignedBox].
+    /// ```none
+    ///          (5)------(1)               Y
+    ///           | \      | \              |
+    ///           |  (4)------(0) MAX       o---X
+    ///           |   |    |   |             \
+    ///      MIN (7)--|---(3)  |              Z
+    ///             \ |      \ |
+    ///              (6)------(2)
+    /// ```
+    pub fn vertices(&self) -> [Vec3; 8] {
+        let min = self.min;
+        let max = self.max;
+        [
+            Vec3::new(max.x, max.y, max.z),
+            Vec3::new(max.x, max.y, min.z),
+            Vec3::new(max.x, min.y, max.z),
+            Vec3::new(max.x, min.y, min.z),
+            Vec3::new(min.x, max.y, max.z),
+            Vec3::new(min.x, max.y, min.z),
+            Vec3::new(min.x, min.y, max.z),
+            Vec3::new(min.x, min.y, min.z),
+        ]
+    }
+    /// Construct an [AxisAlignedBox] given the coordinates of the minimum and maximum corners.
+    pub fn from_min_max(min: Vec3, max: Vec3) -> Result<AxisAlignedBox, PrimitiveError> {
+        if (max - min).min_element() >= 0.0 {
+            Ok(AxisAlignedBox { min, max })
+        } else {
+            Err(PrimitiveError::MinGreaterThanMax)
+        }
+    }
+    /// Construct an [AxisALignedBox] from the origin at the minimum corner, and the extents - the
+    /// dimensions of the box in each axis.
+    pub fn from_extents_origin(
+        extents: Vec3,
+        origin: Vec3,
+    ) -> Result<AxisAlignedBox, PrimitiveError> {
+        if extents.min_element() > 0.0 {
+            Ok(AxisAlignedBox {
+                min: origin,
+                max: extents + origin,
+            })
+        } else {
+            Err(PrimitiveError::NonPositiveExtents)
+        }
+    }
+    /// Computes the AAB that
+    pub fn from_points(points: Vec<Vec3>) -> AxisAlignedBox {
+        let mut max = Vec3::splat(f32::MIN);
+        let mut min = Vec3::splat(f32::MAX);
+        for &point in points.iter() {
+            max = point.max(max);
+            min = point.min(min);
+        }
+        // Unwrap is okay here because min < max
+        AxisAlignedBox::from_min_max(min, max).unwrap()
+    }
+}
+
+/// A frustum is a truncated pyramid that is used to represent the volume of world space that is
+/// visible to the camera.
+#[derive(Copy, Clone, PartialEq, Debug, Reflect)]
+#[reflect_value(PartialEq)]
+pub struct Frustum {
+    planes: [Plane; 6],
+    vertices: [Vec3; 8],
+}
+impl Primitive3d for Frustum {
+    fn outside_plane(&self, plane: Plane) -> bool {
+        for vertex in self.vertices().iter() {
+            if plane.distance_to_point(vertex) <= 0.0 {
+                return false;
+            }
+        }
+        true
+    }
+}
+impl Frustum {
+    fn compute_vertices(camera_position: Mat4, projection_matrix: Mat4) -> [Vec3; 8] {
```

Comment:
Mat4 is pretty large, it's probably better to use a `&Mat4` here

Reply:
Can you comment on the cost of indirection here? I don't have an intuitive sense of the cost of indirection speed vs. memory efficiency.

Reply:
It should be negligible. Like anything the best way to know for sure is to measure through benchmarking/profiling. In glam reference vs value for matrix parameters didn't tend to impact performance because everything was marked as `#[inline]`, so the function call ends up in generated code. If it's not inlined then I suspect that pass by value would be slower.